### PR TITLE
watchdog: wdt_nrfx: return errors, not assertions

### PR DESCRIPTION
**Pull request by JordanYates**
Return an error code if `wdt_feed` is called before `wdt_setup`, instead
of triggering the state assertion inside of `nrfx_wdt_channel_feed`.

Return the documented error if `wdt_install_timeout` is called after
`wdt_setup`, instead of triggering the state assertion inside of
`nrfx_wdt_channel_alloc`.

Don't zero initialise the static data structure, as statics are
explicitly initialised to 0 per the C standard, and checkpatch normally
complains about the pattern.

### CODE REVIEW

File: drivers/watchdog/wdt_nrfx.c [hunk 1]
```diff
@@ -53,6 +54,7 @@ static int wdt_nrf_setup(const struct device *dev, uint8_t options)
 
 	nrfx_wdt_enable(&config->wdt);
 
+	data->enabled = true;
```
Comment (anangl) [inline drivers/watchdog/wdt_nrfx.c:57]: You also need to add `data->enabled = false;` in `wdt_nrf_disable()`.

File: drivers/watchdog/wdt_nrfx.c [hunk 2]
```diff
@@ -141,6 +147,10 @@ static int wdt_nrf_feed(const struct device *dev, int channel_id)
 	if ((channel_id >= data->m_allocated_channels) || (channel_id < 0)) {
 		return -EINVAL;
 	}
+	if (data->enabled == false) {
```
Comment (anangl) [inline drivers/watchdog/wdt_nrfx.c:150]: Why not just `if (!data->enabled) {`?

File: drivers/watchdog/wdt_nrfx.c (hunk 1 above)
Comment (JordanYates) [reply drivers/watchdog/wdt_nrfx.c:57]: done

File: drivers/watchdog/wdt_nrfx.c (hunk 2 above)
Comment (JordanYates) [reply drivers/watchdog/wdt_nrfx.c:151]: There was no other pattern in the file to follow, they are equivalent, and the other checks in the file seem to be explicit
`if (cfg->callback != NULL)`

File: drivers/watchdog/wdt_nrfx.c [hunk 3]
```diff
@@ -17,6 +17,7 @@ struct wdt_nrfx_data {
 	wdt_callback_t m_callbacks[NRF_WDT_CHANNEL_NUMBER];
 	uint32_t m_timeout;
 	uint8_t m_allocated_channels;
+	bool enabled;
```
Comment (kowalewskijan) [inline drivers/watchdog/wdt_nrfx.c:20]: This is non-blocking but to make it common across other wdt drivers I suggest using either `timeout_installed` or `timeout_valid`

File: drivers/watchdog/wdt_nrfx.c [hunk 4]
```diff
@@ -93,6 +96,10 @@ static int wdt_nrf_install_timeout(const struct device *dev,
 	nrfx_err_t err_code;
 	nrfx_wdt_channel_id channel_id;
 
+	if (data->enabled) {
+		return -EBUSY;
```
Comment (kowalewskijan) [inline drivers/watchdog/wdt_nrfx.c:100]: Please add error log here something like: `LOG_ERR("No more timeouts can be installed");` Also I think it should be `-ENOMEM`

**Review by kowalewskijan — CHANGES_REQUESTED** [review]

File: drivers/watchdog/wdt_nrfx.c (hunk 4 above)
Comment (JordanYates) [reply drivers/watchdog/wdt_nrfx.c:100]: `-ENOMEM` is the documented return code:
https://github.com/zephyrproject-rtos/zephyr/blob/0f776cf0bfe85caed8281636354eb72454053624/include/zephyr/drivers/watchdog.h#L208-L209

File: drivers/watchdog/wdt_nrfx.c (hunk 3 above)
Comment (JordanYates) [reply drivers/watchdog/wdt_nrfx.c:20]: Naming does not appear to be consistent across drivers:
https://github.com/zephyrproject-rtos/zephyr/blob/0f776cf0bfe85caed8281636354eb72454053624/drivers/watchdog/wdt_rpi_pico.c#L28

File: drivers/watchdog/wdt_nrfx.c (hunk 4 above)
Comment (JordanYates) [reply drivers/watchdog/wdt_nrfx.c:100]: None of the other errors in this driver result in log messages, I am attempting to keep the changes as small as possible due to the v3.7 RC window.

File: drivers/watchdog/wdt_nrfx.c (hunk 2 above)
Comment (anangl) [reply drivers/watchdog/wdt_nrfx.c:151]: > There was no other bool check pattern in the file to follow

Yes, there was no such pattern, but in `wdt_nrf_install_timeout()` you've added `if (data->enabled)` (which IMO looks much better than `if (data->enabled != false)`, so why not follow that? It's just a detail, I'm not going to insist.

**Review by anangl — DISMISSED** [review]

**Review by anangl — APPROVED** [review]

**Review by jaz1-nordic — APPROVED** [review]

**Pull request merged**